0278-template-only-components.md

```diff
@@ -0,0 +1,249 @@
+RFC. We can therefore (at a later time, in a separate RFC) explore deprecating
+partials in favor of wrapper-free template-only components.
+
+# Detailed design
+
+## API Surface
+
+We should not expose the flag directly as a public API. Instead, we should
+abstract the flag with a "privileged addon" whose only purpose is to enable
+the flag. Applications will enable the flag by installing this addon. This
+will allow for more flexibility in changing the flag's implementation (the
+location, naming, value, or even the existence of it) in the future. From the
+user's perspective, it is the _addon_ that provides thisfunctionality. The
+flag is simply an internal implementation detail.
+
+We have done this before in other cases (such as the legacy view addon during
+the 2.0 transition period), and it has generally worked well.
+
+When landing this feature, it will be entirely opt-in for existing apps, but
+the Ember CLI application blueprint should be updated to include the addon by
+default. At a later time, we should provide another addon that _disables_ the
+flag explicitly (installing both addons would be an install-time error). At
+that time, we will issue a deprecation warning if the flag is *not set*, with
+a message that directs the user to install one of the two addons.
+
+## Single Global "Flag"
+
+The proposed flag will be truly global in scope. That is, setting this flag
+will change the semantics of all template-only components in the entire app,
+even for components that were included by addons.
+
+However, we believe this would not affect any addon components in practice,
+as the predominant pattern for addons to expose components currently
+necessitates a JavaScript class. Addon authors would create the component
+(with or without a JavaScript class) in the `/addon` folder, but exposing
+it for consumption in apps requires creating a corresponding JavaScript class
+in the `/app` folder to "re-export" the component. Therefore, in practice,
+it is not actually possible for addons to have a truely template-only
+component today (something to address in a future RFC).
+
+## Leakage Of `Ember.Component` Semantics
+
+While the primary purpose of this flag is to remove the wrapper element from
+template-only components, there are a few other observable semantics changes
+that comes with it as well.
+
+Currently, template-only components are "backed" by an instance of `Ember.Component`.
+That is, Ember will create an instance of `Ember.Component` and set it as the
+`{{this}}` context for the template.
+
+With the flag enabled, there will be *no* component instance for the template
+and `{{this}}` will be set to `undefined` (or `null`, perhaps). This would
+improve performance for template-only components significantly.
+
+Since there is no JavaScript file for the component, this is only observable
+in a few limited ways:
+
+1. The most noticable artifact is the component's arguments will not be
```

Comment:
How will / should the Ember Inspector behave with template-only components?

cc @teddyzeenny

Reply:
The more general question is how should Ember Inspector work with Glimmer Components and other type of custom components.

I think there are two kinds of components – those that should and should not be shown in the inspector's component tree. For example, `{{render}}`, `{{mount}}`, `{{partial}}` and `{{outlet}}` are all technically implemented as components in the system, but are invisible to the inspector. You could imagine maybe one day things like `{{#if}}` could be implemented as a component as well (though unlikely), but should be invisible to the inspector as well. (These are "virtual" views in old-school Ember.)

(A related question, whether a component should be "visible" in the `Ember.Component` `parentView` (:disappointed:) hierechary.)

I think we need to decide:

1. Should template-only components be visible?
2. Separately, what are the APIs needed to expose a component to the inspector in general.

I think my default position will be that let's make template-only components (and possibly all other custom components) invisible to start while we work out the second question.

Reply:
The inspector currently displays both outlets and components. It builds the view tree by [merging two trees](https://github.com/emberjs/ember-inspector/blob/1c34df9e8e7f521b380d09d6f7d3af15c650cc78/ember_debug/libs/glimmer-tree.js#L116-L122) together: the outlet tree, and the component tree:
1) Builds the outlet tree by walking the `outletState`
2) Builds the component trees by starting with the top component and walking the component tree by getting the child views of each component instance (which are stored as an array on their respective parent instance).
3) Attaches the component sub-trees to their respective parent outlet by matching the controllers of both.
4) The final tree containing the outlets and components is displayed in the View Tree.

With the current approach, template-only components will not be visible as they're neither outlets nor component instances. While certainly not a blocker for the feature, my opinion is that eventually these components should be visible in the inspector as I would expect to see them as a user, same reason why we display outlets which are not backed by component instances. If we can find a way to build template-only component trees and match them to their respective parent outlet/component instance (just like we matched component trees with their parent outlet), then this would work. If not, then I think we would need to change our approach of how we build the view tree.

Reply:
I also think that the template-only components should be visible in the inspector. Until that is done, though, what would happen in the case of a template-only component that invokes a non-template only component? Would the tree of components just end before the template-only component or would it skip over the template-only component?

For clarification:
In, say, application.hbs:
```
{{template-only-component}}
```
In template-only-component.hbs:
```
{{not-template-only-component}}
```

Reply:
@kategengler I think that would depend on the answer to @chancancode's question of "whether a component should be "visible" in the Ember.Component parentView hierechary"? My guess is it would be invisible, similar to "virtual" views, which means the view tree would just skip over the template-only component and display the instance-backed child.

Reply:
Yes. I think for now, they would be invisible by default. I _think_ we'll most likely want to keep the `parentView` tree as-is. In general I believe we want to discourage that style of programming model and use other kind of compositions, like yielding, etc.

It will also likely be a breaking change to expose anything that is _not_ `Ember.Component` as their `parentView`. For example, any code that is walking that tree with a while loop (keep getting the `parentView`) would break, as those custom component might not have the same API. (Or in this case, there is literally no component instance.) Finally, the name (parent**View**) is a misnomer at this point anyway. I think we should probably just deprecate those properties on `Ember.Component`.

Debugging (Ember Inspector) is a different story. We can provide different util methods to walk the tree intended for debugging purpose. Ideally, we should get Ember Inspector to somehow register its presence somehow before the app boots so that we would only need to track that when it is enabled.

Anyhow, that should be discussed separately IMO. Perhaps the [Custom Manager RFC](https://github.com/emberjs/rfcs/pull/213) would be the right place for it, or it could even be further spun-off into another RFC (I would prefer that if we can).

In the short term, I think the status quo would be that these new template-only components will be invisible to the view tree, which is not amazing but not fatal IMO. This is how partials work today (which is probably also not amazing), and the use cases this targets is quite similar.

Reply:
I think invisible (ala partials) is fine, but any components in the tree below the template-only component(s) need to be visible or it would severely hinder the ability to debug.

Reply:
@kategengler correct, that is how it should work. Essentially, both the `parentView` tree and the inspector (they are more or less the same thing today) will skip over anything that is not an `Ember.Component`. Since [my patch have landed in canary](https://github.com/emberjs/ember.js/pull/15974), we should be able to boot an app and see if this is breaking anything in the inspector now.

Reply:
I did just that before adding my comment earlier today: In Ember Inspector, partials are just invisible, but unfortunately, template-only components cut off the tree. Since the implementation is already in canary, should I go ahead and file an issue on the inspector for this?

Reply:
@kategengler yes please! Thank you for checking, I'll look into it!

Reply:
You can also file that bug against Ember, as I suspect it's a problem on my end. Either way, as long as you mention or assign it to me I can take a look.